### PR TITLE
HPR-1174: Add `hcp_packer_run_task` resource and data source

### PR DESCRIPTION
**Pull request by aidan-mundy**
### :hammer_and_wrench: Description

- Adds new Resource: `hcp_packer_run_task`
- Adds new Data Source `hcp_packer_run_task`
- Minor code duplication cleanup for location getting/setting on Packer resources

**NOTE: I'm planning to add a "guide" page with documentation for integration with the TFE provider. I'll likely do this Wednesday morning after the holiday.**

### :building_construction: Acceptance tests

- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch?

Output from acceptance testing:

New tests
```
$ make testacc TESTARGS='-run=Test.*Packer.*Run.* -test.v'
==> Checking that code complies with gofmt requirements...
golangci-lint run --config ./golangci-config.yml 
TF_ACC=1 go test ./internal/... -v -run=Test.*Packer.*Ru.* -test.v -timeout 360m -parallel=10
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/clients    0.420s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/consul     0.191s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/input      0.230s [no tests to run]
=== RUN   TestAcc_dataSourcePackerRunTask
--- PASS: TestAcc_dataSourcePackerRunTask (9.43s)
=== RUN   TestAccPackerRunTask
--- PASS: TestAccPackerRunTask (17.28s)
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/provider   27.400s
```

All Packer tests
```
$ make testacc TESTARGS='-run=Test.*Packer.* -test.v'
==> Checking that code complies with gofmt requirements...
golangci-lint run --config ./golangci-config.yml 
TF_ACC=1 go test ./internal/... -v -run=Test.*Packer.* -test.v -timeout 360m -parallel=10
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/clients    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/consul     (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/input      (cached) [no tests to run]
=== RUN   TestAcc_dataSourcePackerBucketNames
--- PASS: TestAcc_dataSourcePackerBucketNames (17.01s)
=== RUN   TestAcc_dataSourcePacker
--- PASS: TestAcc_dataSourcePacker (5.75s)
=== RUN   TestAcc_dataSourcePacker_revokedIteration
--- PASS: TestAcc_dataSourcePacker_revokedIteration (10.82s)
=== RUN   TestAcc_dataSourcePackerImage
--- PASS: TestAcc_dataSourcePackerImage (7.25s)
=== RUN   TestAcc_dataSourcePackerImage_revokedIteration
--- PASS: TestAcc_dataSourcePackerImage_revokedIteration (5.57s)
=== RUN   TestAcc_dataSourcePackerImage_emptyChannel
--- PASS: TestAcc_dataSourcePackerImage_emptyChannel (1.37s)
=== RUN   TestAcc_dataSourcePackerImage_channelAndIterationIDReject
--- PASS: TestAcc_dataSourcePackerImage_channelAndIterationIDReject (3.12s)
=== RUN   TestAcc_dataSourcePackerImage_channelAccept
--- PASS: TestAcc_dataSourcePackerImage_channelAccept (6.21s)
=== RUN   TestAcc_dataSourcePackerIteration
--- PASS: TestAcc_dataSourcePackerIteration (5.54s)
=== RUN   TestAcc_dataSourcePackerIteration_revokedIteration
--- PASS: TestAcc_dataSourcePackerIteration_revokedIteration (10.96s)
=== RUN   TestAcc_dataSourcePackerRunTask
--- PASS: TestAcc_dataSourcePackerRunTask (8.32s)
=== RUN   TestAccPackerChannelAssignment_SimpleSetUnset
--- PASS: TestAccPackerChannelAssignment_SimpleSetUnset (13.10s)
=== RUN   TestAccPackerChannelAssignment_AssignLatest
--- PASS: TestAccPackerChannelAssignment_AssignLatest (15.88s)
=== RUN   TestAccPackerChannelAssignment_InvalidInputs
--- PASS: TestAccPackerChannelAssignment_InvalidInputs (6.18s)
=== RUN   TestAccPackerChannelAssignment_CreateFailsWhenPreassigned
--- PASS: TestAccPackerChannelAssignment_CreateFailsWhenPreassigned (8.55s)
=== RUN   TestAccPackerChannelAssignment_HCPManagedChannelErrors
--- PASS: TestAccPackerChannelAssignment_HCPManagedChannelErrors (3.62s)
=== RUN   TestAccPackerChannelAssignment_EnforceNull
--- PASS: TestAccPackerChannelAssignment_EnforceNull (29.88s)
=== RUN   TestAccPackerChannel
--- PASS: TestAccPackerChannel (6.60s)
=== RUN   TestAccPackerChannel_AssignedIteration
--- PASS: TestAccPackerChannel_AssignedIteration (8.39s)
=== RUN   TestAccPackerChannel_UpdateAssignedIteration
--- PASS: TestAccPackerChannel_UpdateAssignedIteration (12.95s)
=== RUN   TestAccPackerChannel_UpdateAssignedIterationWithFingerprint
--- PASS: TestAccPackerChannel_UpdateAssignedIterationWithFingerprint (6.10s)
=== RUN   TestAccPackerRunTask
--- PASS: TestAccPackerRunTask (17.75s)
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/provider   211.345s
```

### CODE REVIEW

File: internal/provider/data_source_packer_image_iteration.go [hunk 1]
```diff
@@ -207,17 +198,6 @@ func dataSourcePackerImageIterationRead(ctx context.Context, d *schema.ResourceD
 	return nil
 }
 
-func setLocationData(d *schema.ResourceData, loc *sharedmodels.HashicorpCloudLocationLocation) error {
```
Comment (sylviamoss) [inline internal/provider/data_source_packer_image_iteration.go:210]: I guess this was resetting those fields unnecessarily, right?

File: internal/provider/data_source_packer_run_task.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func dataSourcePackerRunTaskRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client)
+	loc, err := getAndUpdateLocationResourceData(d, client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(loc.ProjectID)
```
Comment (sylviamoss) [inline internal/provider/data_source_packer_run_task.go:63]: Is this really needed? This is not being done in the other resources after you updated the code.

File: internal/provider/resource_packer_run_task.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	loc, err := getAndUpdateLocationResourceData(d, client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp, err := clients.GetRunTask(ctx, client, loc)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(loc.ProjectID)
```
Comment (sylviamoss) [inline internal/provider/resource_packer_run_task.go:80]: Same here!

File: internal/provider/resource_packer_run_task.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		if err := d.Set("hmac_key", resp.HmacKey); err != nil {
+			return diag.FromErr(err)
+		}
+		// Always set `regenerate_hmac` to false so that it can be set to false by
+		// the user without generating a diff.
+		if err := d.Set("regenerate_hmac", false); err != nil {
+			return diag.FromErr(err)
```
Comment (sylviamoss) [inline internal/provider/resource_packer_run_task.go:100]: What happens if you don't do this?

File: internal/provider/data_source_packer_image_iteration.go (hunk 1 above)
Comment (aidan-mundy) [reply internal/provider/data_source_packer_image_iteration.go:210]: Not necessarily. I replaced this with functionally similar code in `location.go`

File: internal/provider/resource_packer_run_task.go (hunk 4 above)
Comment (aidan-mundy) [reply internal/provider/resource_packer_run_task.go:100]: Since `regenerate_hmac` is intended to be a one-time operation, this automatically toggles it back to false in the state.

- With this code
  - A plan will always be generated when `regenerate_hmac = true`. 
    - Plans show that `hmac_key` will regenerate
    - Plans with `regenerate_hmac = true` will always show that the state contains `regenerate_hmac = false`, even if applied multiple times in a row
  - A plan will never be generated when `regenerate_hmac = false`. 
- Without this code
  - A plan will always be generated when `regenerate_hmac = true`. 
    - Plans show that `hmac_key` will regenerate
  - A plan _may_ be generated when `regenerate_hmac = false` _if_ the config contained `regenerate_hmac = true` during the last apply

TL;DR: This code removes the (effectively) no-op plan when applying with `regenerate_hmac = false` after previously applying with `regenerate_hmac = true`

File: internal/provider/data_source_packer_run_task.go (hunk 2 above)
Comment (sylviamoss) [reply internal/provider/data_source_packer_run_task.go:63]: True!

File: internal/provider/resource_packer_run_task.go (hunk 4 above)
Comment (sylviamoss) [reply internal/provider/resource_packer_run_task.go:104]: Nice! Thanks for explaining!

**Review by sylviamoss — APPROVED** [review]
Nice! 👍🏼
